src/Conversion/ONNXToKrnl/RNN/RNNBase.hpp

```diff
@@ -0,0 +1,131 @@
+    auto loc = op->getLoc();
+
+    I inputPack;
+    O outputPack;
+    S state;
+
+    std::tie(inputPack, outputPack) =
+        getInputOutputPack<RNNOp, I, O>(op, operands);
+
+    if (hasNoOutput<O>(outputPack)) {
+      rewriter.eraseOp(op);
+      return success();
+    }
+
+    state = allocAndInitializeStates<I, O, S>(
+        rewriter, loc, op, inputPack, outputPack);
+    A activationForward, activationReverse;
+    std::tie(activationForward, activationReverse) =
+        getActivationPack<RNNOp, A>(op);
+
+    Value X = inputPack.X;
+    int sequenceLengthDim = X.getType().cast<ShapedType>().getShape()[0];
+    auto direction = llvm::dyn_cast<RNNOp>(op).direction();
+
+    if (direction == FORWARD || direction == BIDIRECTIONAL) {
+      BuildKrnlLoop sequenceLoops(rewriter, loc, 1);
+      sequenceLoops.createDefineAndOptimizeOp();
+      sequenceLoops.pushBounds(0, sequenceLengthDim);
+      sequenceLoops.createIterateOp();
+
+      auto ipSequenceLoops = rewriter.saveInsertionPoint();
+      rewriter.setInsertionPointToStart(sequenceLoops.getIterateBlock());
+      {
+        Value numDirectionIV =
+            emitConstantOp(rewriter, loc, rewriter.getIndexType(), 0);
+        Value sequenceLengthIV = sequenceLoops.getInductionVar(0);
+        // Emit calculation for one RNN step.
+        calculateState<RNNOp, I, S>(rewriter, loc, op, numDirectionIV,
+            sequenceLengthIV, inputPack, state, activationForward);
+      }
+      rewriter.restoreInsertionPoint(ipSequenceLoops);
+    }
+
+    if (direction == REVERSE || direction == BIDIRECTIONAL) {
+      BuildKrnlLoop sequenceLoops(rewriter, loc, 1);
+      sequenceLoops.createDefineAndOptimizeOp();
+      sequenceLoops.pushBounds(0, sequenceLengthDim);
+      sequenceLoops.createIterateOp();
+
+      auto ipSequenceLoops = rewriter.saveInsertionPoint();
+      rewriter.setInsertionPointToStart(sequenceLoops.getIterateBlock());
+      {
+        AffineMap reverseIVMap = AffineMap::get(1, 1,
+            rewriter.getAffineSymbolExpr(0) - rewriter.getAffineDimExpr(0) - 1);
+
+        Value numDirectionIV = emitConstantOp(rewriter, loc,
+            rewriter.getIndexType(), (direction == REVERSE) ? 0 : 1);
+        Value reverseSequenceLengthIV =
+            rewriter.create<AffineApplyOp>(loc, reverseIVMap,
+                ValueRange(std::vector<Value>{sequenceLoops.getInductionVar(0),
+                    emitConstantOp(rewriter, loc, rewriter.getIndexType(),
+                        sequenceLengthDim)}));
+        // Emit calculation for one RNN step.
+        calculateState<RNNOp, I, S>(rewriter, loc, op, numDirectionIV,
+            reverseSequenceLengthIV, inputPack, state, activationReverse);
+      }
+      rewriter.restoreInsertionPoint(ipSequenceLoops);
+    }
+
+    std::vector<Value> outputs;
+    stateToOutput<RNNOp, S, O>(state, outputPack, outputs);
+    rewriter.replaceOp(op, llvm::makeArrayRef(outputs));
```

Comment:
Curious, do we actually need llvm::makeArrayRef here? i.e., if you don't make an explicit type cast, does it happen implicitly nevertheless?